contracts/Raffles.cdc

```diff
@@ -0,0 +1,142 @@
+import "MetadataViews"
+
+pub contract Raffles {
+    pub let ManagerStoragePath: StoragePath
+    pub let ManagerPublicPath: PublicPath
+
+    pub event RaffleCreated(address: Address, raffleID: UInt64, sourceType: Type)
+    pub event RaffleDrawn(address: Address, raffleID: UInt64, sourceType: Type, index: Int, value: String)
+
+    pub struct Details {
+        pub let start: UInt64?
+        pub let end: UInt64?
+        pub let display: MetadataViews.Display
+
+        init(
+            _ start: UInt64?,
+            _ end: UInt64,
+            _ display: MetadataViews.Display
+            
+        ) {
+            self.start = start
+            self.end = end
+            self.display = display
+        }
+    }
+
+    pub resource interface RafflePublic {
+        pub fun getEntryAt(index: Int): AnyStruct
+    }
+
+    pub resource Raffle: RafflePublic, MetadataViews.Resolver {
+        pub let source: @{RaffleSource}
+        pub let details: Details
+
+        pub fun draw(): Int {
+            return self.source.draw()
+        }
+
+        pub fun getEntryAt(index: Int): AnyStruct {
+            return self.source.getEntryAt(index: index)
+        }
+
+        pub fun getViews(): [Type] {
+            return [
+                Type<MetadataViews.Display>()
+            ]
+        }
+
+        pub fun resolveView(_ type: Type): AnyStruct? {
+            switch type {
+                case Type<MetadataViews.Display>():
+                    return self.details.display
+            }
+
+            return nil
+        }
+
+        init(
+            source: @{RaffleSource},
+            start: UInt64?,
+            end: UInt64,
+            display: MetadataViews.Display
+        ) {
+            self.source <- source
+            self.details = Details(start, end, display)
+        }
+
+        destroy() {
+            destroy self.source
+        }
+    }
+
+    pub resource interface RaffleSource {
+        pub fun draw(): Int
+        pub fun getEntryAt(index: Int): AnyStruct
+    }
+
+    pub resource AddressRaffleSource: RaffleSource {
+        pub let addresses: [Address]
+
+        pub fun draw(): Int {
+            let r = revertibleRandom()
+            return Int(UInt64(self.addresses.length) % r)
+        }
+
+        pub fun getEntryAt(index: Int): AnyStruct {
```

Comment:
This fn will return a Address, right?
Do we plan to change to return something else or we could add here:

`pub fun getEntryAt(index: Int): Address {`

Reply:
It might return an address in one implementation, but if we enforce it at the interface, only addresses would be allowed to be used as raffle sources which isn't going to be the case. We've already seen twitter raffles using on-chain randomness, for example